### ISSUE TITLE
What about: mixed-size atomic accesses

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#345]
So it looks like mixed-size-accesses [made the round on twitter](https://twitter.com/maybewaffle/status/1542570724489134080) again recently which got me thinking about them. https://github.com/rust-lang/rust/pull/97516 carefully chose not to talk about them (and anyway it's not clear that is the right place to go into more detail about them). So what *do* we want to do with them in Rust?

Some facts:
- In C++, you cannot convert a `&uint16_t` into an reference to an array "because no such array exists at that location in memory"; they insist that memory is strongly typed. This means they don't even have to talk about mixed-size accesses. It also means they are ignoring a large fraction of the programs out there but I guess they are fine with that. We are not. ;)
- Apparently the x86 manual says you "should" not do this: "Software should access semaphores (shared memory used for signalling between multiple processors) using identical addresses and operand lengths." It is unclear what "should" means (or what anything else here really means, operationally speaking...)
- In Rust, it is pretty much established that you can safely turn a `&mut u16` into a `&mut [u8; 2]`. So you can do something where you start with a `&mut AtomicU16`, do some atomic things to it, then use the above conversion and get a `&mut AtomicU8` and do atomic things with *that* -- a bona fide mixed-size atomic access.

    *However*, this means that there is a happens-before edge between all the 16-bit accesses and the 8-bit accesses. Having the `&mut` means that there is synchronization. I *hope* this means not even Intel disagrees with this, but since literally none of the words in that sentence is defined, who knows.

So... it seems like the most restrictive thing we can say, without disallowing code that you can already write entirely safely using bytemuck, is that
- it is allowed to do differently-sized atomic accesses to the same location over time,
- **but only** if any two not-perfectly-overlapping accesses are completely synchronized through other means (i.e., it is not these accesses themselves that add a happens-before edge, there already exists a happens-before edge through other accesses).
- Any other kind of mixed-size access is UB.

Cc @workingjubilee @m-ou-se @Amanieu @cbeuw @thomcc

**Comment by m-ou-se** [comment]
That all matches my understanding as well.

(It's part of the reason why I added `Atomic*::from_mut`.)

> However, this means that there is a happens-before edge between all the 16-bit accesses and the 8-bit accesses. Having the &mut means that there is synchronization. I hope this means not even Intel disagrees with this, but since literally none of the words in that sentence is defined, who knows.

If they would disagree with that, that'd basically imply that after using some memory for an atomic operation, you can never re-use that memory again. E.g. deallocating a Box<AtomicU16> would be unsafe, and so would be a stack-allocated AtomicU16 that goes out of scope.

They don't say it very clearly, but I don't see how their no-mixed-sizes rule can apply to anything other than atomic operations on the same memory that race with each other.

> It is unclear what "should" means (or what anything else here really means, operationally speaking...)

Yeah, it could very well turn out that "should" just means "for performance", and that it has nothing to do with correctness. They're not very clear.

> So... it seems like the most restrictive thing we can say [..]

That seems like exactly the right thing to say, and matches what you can do in safe Rust (if we include the unstable `Atomic*::from_mut`).

I don't think it's impossible that this might be less restrictive in the future, if we find more reasons to believe that racing mixed-size atomic operations will work on all platforms.

> In C++, you cannot convert a &uint16_t into an reference to an array

Converting `uint16_t*` to a `char*` however is fine, e.g. to memset or memcpy into a uint16_t or struct, etc.

In C++20, you can also have a `struct X { int a; int b; }` and create an `std::atomic_ref<X>` first, and an `std::atomic_ref<int>` to one of the fields later.

In [atomics.ref.generic#general-3](https://eel.is/c++draft/atomics.ref.generic#general-3), they clearly specify mixed-size accesses in the same way as us:

> The lifetime ([basic.life]) of an object referenced by `*ptr` shall exceed the lifetime of all `atomic_­ref`s that reference the object. While any `atomic_­ref` instances exist that reference the `*ptr` object, all accesses to that object shall exclusively occur through those `atomic_­ref` instances.
**No subobject of the object referenced by `atomic_­ref` shall be concurrently referenced by any other `atomic_­ref` object.**

**Comment by Amanieu** [comment]
ARM's [memory model](https://developer.arm.com/documentation/ddi0487/latest) (in the section: The AArch64 Application Level Memory Model) seems to fully support mixed-sized atomic accesses.

**Comment by RalfJung** [comment]
> If they would disagree with that, that'd basically imply that after using some memory for an atomic operation, you can never re-use that memory again. E.g. deallocating a Box would be unsafe, and so would be a stack-allocated AtomicU16 that goes out of scope.

Ah, good point. I had forgotten that hardware memory models do not have provenance. :joy: 

> That seems like exactly the right thing to say, and matches what you can do in safe Rust (if we include the unstable Atomic*::from_mut).

.. and include bytemuck

> Converting uint16_t* to a char* however is fine, e.g. to memset or memcpy into a uint16_t or struct, etc.

Isn't that a C thing? Though C++ might have something similar with `std::byte`.
But anyway that's a non-atomic type.

> In C++20, you can also have a struct X { int a; int b; } and create an std::atomic_ref<X> first, and an std::atomic_ref<int> to one of the fields later.

Oh, good point. So in some sense this is actually already all covered by https://github.com/rust-lang/rust/pull/97516.

**Comment by bjorn3** [comment]
> > If they would disagree with that, that'd basically imply that after using some memory for an atomic operation, you can never re-use that memory again. E.g. deallocating a Box would be unsafe, and so would be a stack-allocated AtomicU16 that goes out of scope.
> 
> Ah, good point. I had forgotten that hardware memory models do not have provenance. joy

When reusing memory it is undef, right? Furthermore deallocation requires some kind of synchronization with every thread that has ever accessed it using atomic operations. Together I would assume this is enough to provide consistency by "resetting" the state witnessing that it was accessed using atomic operations of a different size.

**Comment by m-ou-se** [comment]
> Though C++ might have something similar with `std::byte`.

C++ allows aliasing through `char`, `unsigned char` and `std::byte`: https://wg21.link/basic.lval#11.3

> std::atomic_ref

To add to my previous comment: one of the reasons why C++'s atomic_ref doesn't allow mixed size / overlapping operations, is that it supports objects of any size. If it gets too big for native atomic instructions, it uses a mutex instead, which is probably stored in some kind of global table indexed by the address of the object. It's not completely clear whether it's necessary to be as restrictive when limited to only natively supported atomic operations, like in Rust.

**Comment by chorman0773** [comment]
> Apparently the x86 manual says you "should" not do this: "Software should access semaphores (shared memory used for signalling between multiple processors) using identical addresses and operand lengths." It is unclear what "should" means (or what anything else here really means, operationally speaking...)

I thik we can interpet "should" as "It's undefined, by spec, though it works in practice b/c some important people rely on it, but please don't, we want to do fast things".

**Comment by cbeuw** [comment]
Regarding x86, I got [the following](https://lists.isocpp.org/std-proposals/2023/04/6237.php) from @thiagomacieira which is very helpful

> Even on Intel processors (note: I work for Intel and this is an area I am very familiar
with) you *can* mix different-sized operations and still retain atomicity,
provided you obey some rules:
>
> * never cross a cacheline boundary (preferably, align naturally)
> * use operations of 16 bytes or less (ideally: use only 1-uop operations)
>   * for all current in-market processors and I believe this applies to AMD
>   * for all current P-core processors: 32- and 64-byte accesses are also fine
>
> Note: I don't know how valid this is for the new RAO instructions, but they
should be ok for CMPccXADD.
>
> In fact, I've seen a few codebases that do use the fact that they can mix two atomic_uint32_t with an overlapping atomic_uint64_t, at least when it comes to interfacing with the Linux kernel 32-bit futex support, see
> https://codebrowser.dev/glibc/glibc/nptl/sem_waitcommon.c.html#do_futex_wait

**Comment by RalfJung** [comment]
> use operations of 16 bytes or less (ideally: use only 1-uop operations) 

So this means there are atomic 256bit accesses but doing size mixing with those is a problem? (Not sure what "1-uop operations are.)

**Comment by chorman0773** [comment]
On Mon, Apr 10, 2023 at 13:18 Ralf Jung ***@***.***> wrote:

> use operations of 16 bytes or less (ideally: use only 1-uop operations)
>
> So this means there are atomic 256bit accesses but doing size mixing with
> those is a problem? (Not sure what "1-uop operations are.)
>
1-uop operations are operations that compile down to an instruction that
does a single microinstruction. I can't remember if sse movs are 1-uop, but
this would be all of the scalar instructions for sure.

**Comment by thiagomacieira** [comment]
SSE loads and stores are atomic. AVX 256- and 512-bit loads and stores are atomic on P-core processors, but not E-core (the 256-bit operation is cracked into two 128-bit operations and therefore not atomic).

There s no RMW SIMD. The best you get is a merge-store and I'm confident that's atomic on P-core, but I doubt so on E-core. Therefore, SIMD atomics are very limited if you can only use them for loads and stores. The most useful thing that this could be done for is to load 16 bytes atomically, and use CMPXCHG16B to store, but it's still limited and somewhat slow due to the transfer between register files. seqlocks are more flexible.

**Comment by tmandry** [comment]
My reading of the [RISC-V spec](https://github.com/riscv/riscv-isa-manual/releases/download/20240411/unpriv-isa-asciidoc.pdf) says that mixed-size atomics _are_ well-defined if they are aligned, both in terms of their atomicity and in terms of their ordering with respect to other load/store operations in the same thread.

Atomicity:

> All memory operations are single-copy atomic: they can never be observed in a partially complete state.

Ordering (emphasis mine):

> The global memory order for any given execution of a program respects some but not all of each hart’s program order. The subset of program order that must be respected by the global memory order is known as preserved program order.
> 
> The complete definition of preserved program order is as follows (and note that AMOs are simultaneously both loads and stores): memory operation a precedes memory operation b in preserved program order (and hence also in the global memory order) if a precedes b in program order, a and b both access regular main memory (rather than I/O regions), and any of the following hold:
> 
> • **Overlapping-Address Orderings:**
> 1. b is a store, and a and b access **overlapping memory addresses**
> 2. a and b are loads, **x is a byte read by both a and b**, there is no store to x between a and b in program order, and a and b return values for x written by different memory operations
> 3. a is generated by an AMO or SC instruction, b is a load, and **b returns a value written by a**

_Context: Here, AMO is an atomic modify operation such as fetch_add or fetch_xor, and "hart" is a hardware thread. RISC-V does not implement CAS directly but uses paired LR (load-reserved) and SC (store-conditional) instructions, which can be used to implement CAS._

Note that there is a discrepancy described in section A.7.1 between their axiomatic (normative) and operational memory models, regarding mixed-size load and store operations. The examples provided all look like edge cases to me and I do not believe they have any bearing on guarantees provided by the C++ atomic model (even if the model were extended to cover mixed-size atomics). As long as you rely only on atomics being atomic and on acquire/release (or stronger) semantics providing happens-before relationships, you could not write code that depends on this discrepancy being resolved one way or another.

**Comment by RalfJung** [comment]
There are plenty of hardware models that allow mixed-size atomics. That's entirely useless for purpose though; Rust needs a *language-level* memory model and for those, so far I don't know of any proposal that would permit mixed-size atomics. Hardware-level memory models disallow many optimizations we'd want to do (such as GVN, hoisting code, or algebraic identities that remove syntactic dependencies like `y-y → 0`).

**Comment by zachs18** [comment]
Yes, Rust cannot just directly use a hardware memory model, and I do not think tmandry was suggesting it should, but it is still necessary to discuss hardware memory models to determine whether or not a possible future Rust memory model allowing mixed-size atomics is even implementable on those platforms.

**Comment by tmandry** [comment]
Don't worry! I promise to forever believe in the Rust abstract machine.

I also think we have a precedent for "poking holes" in Rust's abstract machine by saying the behavior of certain operations is "whatever the underlying hardware does – within (important) constraints" – correct me if I'm wrong, but floating point comes to mind.

Currently, I don't see any insurmountable obstacles to doing that for mixed-size atomic ops. It's very unlikely that LLVM takes advantage of the UB given that actual C/C++ code in the wild makes use of them, and so far I can't imagine a use case that allows us to do an important optimization (at least, the importance of such an optimization would likely be smaller than the importance of using mixed-size atomics).

Examples of usage in the wild:

1. [glibc semaphores](https://codebrowser.dev/glibc/glibc/nptl/sem_waitcommon.c.html#do_futex_wait), as shared by @thiagomacieira above
2. They [seem to be the basis](https://stackoverflow.com/a/53219577/676226) of the refusal to support 64-bit futexes in Linux
3. A [blocking rw-lock](https://cs.opensource.google/fuchsia/fuchsia/+/main:zircon/kernel/include/kernel/brwlock.h;l=263;drc=6bc94dd044cecfd7eb47e477cd8d7e250c8b765e) in the Fuchsia kernel

While we could tell everyone they should go down to assembly for this, it would be much nicer – and probably more optimization-friendly – if we could let them use Rust atomics.

**Comment by tmandry** [comment]
Yeah, I didn't mean to suggest this was trivial, and your comment makes me realize that "poking a hole" is definitely the wrong framing.

I would very much like it if we could extend the memory model (possibly in target-specific ways) to support these operations, but I don't know how to do that and I'm sure some smart people have tried. On the bright side, we may have ourselves a great idea for a PhD thesis...

**Comment by comex** [comment]
Does anyone here know anyone at Intel that might be able to start a process that leads to Intel actually clarifying their manual?

@thiagomacieira's informal clarification to the C++ WG is a starting point, but it's informal, and even on its own terms it's more like a description of current processor behavior than a guarantee of future processor behavior.

I would argue the current manual is not just too weak but actually self-contradictory.  As previously mentioned, in the Intel® 64 and IA-32 Architectures Software Developer’s Manual, Vol. 3A, 10.1.2.2, there is this clause:

> Software should access semaphores (shared memory used for signaling between multiple processors) using identical addresses and operand lengths. For example, if one processor accesses a semaphore using a word access,
other processors should not access the semaphore using a byte access.

But in the very same section are these other clauses:

> A locked instruction is guaranteed to lock only the area of memory defined by the destination operand, but may be
interpreted by the system as a lock for a larger memory area.

> The integrity of a bus lock is not affected by the alignment of the memory field. The LOCK semantics are followed
for as many bus cycles as necessary to update the entire operand.
> [..]
> Locked operations are atomic with respect to all other memory operations and all externally visible events. Only
instruction fetch and page table accesses can pass locked instructions. Locked instructions can be used to synchronize data written by one processor and read by another processor.

Is the "should access semaphores" clause supposed to be an exception to some part of these guarantees?  If so, which?  (And does "identical addresses" mean virtual or physical addresses?)

Note that earlier, a "locked instruction" is defined as one of the following instructions with a LOCK prefix:

> - The bit test and modify instructions (BTS, BTR, and BTC).
> - The exchange instructions (XADD, CMPXCHG, CMPXCHG8B, and CMPXCHG16B).
> - The LOCK prefix is automatically assumed for XCHG instruction.
> - The following single-operand arithmetic and logical instructions: INC, DEC, NOT, and NEG.
> - The following two-operand arithmetic and logical instructions: ADD, ADC, SUB, SBB, AND, OR, and XOR.

There is a separate section, 10.1.1, that handles atomicity guarantees for regular load and store operations.  Again there is no explicit restriction on access sizes.  I will just quote the whole thing:

> The Intel486 processor (and newer processors since) guarantees that the following basic memory operations will always be carried out atomically:
> - Reading or writing a byte.
> - Reading or writing a word aligned on a 16-bit boundary.
> - Reading or writing a doubleword aligned on a 32-bit boundary.
>
> The Pentium processor (and newer processors since) guarantees that the following additional memory operations will always be carried out atomically:
> - Reading or writing a quadword aligned on a 64-bit boundary.
> - 16-bit accesses to uncached memory locations that fit within a 32-bit data bus.
>
> The P6 family processors (and newer processors since) guarantee that the following additional memory operation will always be carried out atomically:
> - Unaligned 16-, 32-, and 64-bit accesses to cached memory that fit within a cache line.
>
> Processors that enumerate support for Intel® AVX (by setting the feature flag CPUID.01H:ECX.AVX[bit 28]) guarantee that the 16-byte memory operations performed by the following instructions will always be carried out atomically:
> - MOVAPD, MOVAPS, and MOVDQA.
> - VMOVAPD, VMOVAPS, and VMOVDQA when encoded with VEX.128.
> - VMOVAPD, VMOVAPS, VMOVDQA32, and VMOVDQA64 when encoded with EVEX.128 and k0 (masking disabled).
(Note that these instructions require the linear addresses of their memory operands to be 16-byte aligned.)
>
> Accesses to cacheable memory that are split across cache lines and page boundaries are not guaranteed to be atomic […snip details...].
>
> Except as noted above, an x87 instruction or an SSE instruction that accesses data larger than a quadword may be implemented using multiple memory accesses. [...snip wording about faults...]

At least this provides some formal backing for the "1-uop instruction" rule.  The multi-uop instructions are presumably not in the above list.

Honestly, given the longstanding assumptions made by software like Linux, I suspect Intel should just strike the entire "should access semaphores" paragraph and leave the rest as is.  But is there any way we can get them to investigate actually doing so?

**Comment by RalfJung** [comment]
I also wonder what the requirements mentioned [above](https://github.com/rust-lang/unsafe-code-guidelines/issues/345#issuecomment-1499421708) mean for sharing memory with untrusted code. I can make sure that my accesses never cross a cacheline boundary, but if other code concurrently accesses the same memory and uses operations that do span multiple cache lines, or that take multiple uOPs... we'd need a guarantee that only that other process can see strange things happen, but everything will remain sane on my side.

**Comment by thiagomacieira** [comment]
Those are architectural questions that you can't enforce. The answer to what happens when you do atomic and non-atomic accesses to the same memory will depend on the CPU and architecture in question. Given how low-level this is and how critical to performance it is, CPU vendors are unlikely to enforce a specific behaviour and may choose to take different approaches depending on generation too.

Even for x86 I don't want to voice what would happen if you did `LOCK ADD` in one place and plain `ADD` somewhere else.

There will be some guarantees, like an atomic xchg or cmpxchg still being atomic, but the data race that is happening with the other thread simply can't guarantee anything about the state prior to or after the atomic operation completing.

**Comment by chorman0773** [comment]
Oh, and for fun, I believe it can depend on Memory Type as well.
You can't do acq/rel ops (at least, the same say) on WC, WT, or UC memory like you can on WB, but you can do relaxed ops* with WC type.
(Don't worry about what the acronyms mean exactly, the important thing is there are 4 types of memory and memory accesses in terms of writes - Write Back, Write Combining, Write Through, and Uncached. Write Back is the default and the others mostly require OS Configuration, though Write Combining is available through movnt and a number of far more recent extensions).

**Comment by thiagomacieira** [comment]
Normal userspace applications will never see anything except Write Back. All the other types are only usually going to be used by device drivers or similar low-level programming. As a language, Rust will likely interact with that (depending on the mood of kernel developers) but it need not be part of the standard language behaviour. Such code will want to get the hardware and device behaviour, whatever it is.

**Comment by chorman0773** [comment]
Well, as mentioned, non-temporal operations use WC access type on WB Memory.

**Comment by asahilina** [comment]
For what it's worth, GPU drivers often map GPU memory as WC (when the expected usage is upload-only), and this memory is mapped to userspace applications via standard graphics APIs. So normal userspace applications do see at least WC memory with some regularity.

**Comment by RalfJung** [comment]
Rust-accessible memory pretty much has to be "normal" write back memory, everything else is entirely unsupported for now. Let's not derail this thread by trying to support it. :)

> There will be some guarantees, like an atomic xchg or cmpxchg still being atomic, but the data race that is happening with the other thread simply can't guarantee anything about the state prior to or after the atomic operation completing.

It's okay if the memory contents get mangled a bit, but the thread that did the atomic operation must behave completely normal in this situation or it becomes impossible to share memory with untrusted code -- which cannot be what anyone wants.

> If it were actually impossible to access untrusted memory safely, no secure OS could exist that ever shares memory with userspace (which is possible, but would have lots of overhead). So regardless of what the Intel documentation says, either this is possible and doing what e.g. copy_from_user on Linux does is safe, or all OSes are broken and we have bigger problems than Rust guarantees... as comex said, the answer is probably just that the Intel docs are unclear and too restrictive.

Yeah, that's what I was getting at. Though I guess specifically for the Linux case, the trusted code doesn't use arbitrary atomic operations, it only does the equivalent of atomic loads and stores? Or are there situations where the kernel will to an RMW operation on userspace memory?

**Comment by chorman0773** [comment]
> Rust-accessible memory pretty much has to be "normal" write back memory, everything else is entirely unsupported for now. Let's not derail this thread by trying to support it.

I mean, other than maybe this (I'm not sure exactly how mixing memory types and sizes will be handled), WC only needs a slightly different correspondence relation (and slightly different release codegen in particular) to be implemented (as far as I'm aware, that's a choice of the implementation) to be correct for non-atomic and relaxed ops. I'm actually looking at the 2022 atomic instructions with interest (might produce better optimized code than using lock prefixes in code that uses alot of Relaxed).

**Comment by thiagomacieira** [comment]
The kernel may do an RMW operation on userspace memory if it wants to, but the application must assume that is not atomic. I can't think of an example now where it would do that, but I also don't think it's prohibited, because I don't see a problem at all. You *can* mix atomic and non-atomic operations on the same cacheline, on the same datum, as well as overlapping data.

I can't speak about other architectures, but on Intel, this is all coordinated by the cache coherency protocol and thus the locking is done per cacheline. Any 1-uop operations that stay within the bounds of the cacheline will be atomic, so there's no danger of sharing memory between trusted and untrusted code. Operations straddling a cacheline boundary will either be split into two intra-cacheline operations if the LOCK prefix is absent, or will lock the entire bus if it is present (and will generate a splitlock exception for the kernel to handle, since Ice Lake).

In summary:
* aligned operations, whether atomic or not, will not suffer tearing
   * all aligned operations up to 8 bytes are and have always been 1-uop
   * 16-byte loads and store will not tear on any AVX-enabled processor
   * 32- and 64-byte aligned operations are not guaranteed anywhere (but are tear-free on all current Intel P-core processors)
* aligned atomics are atomic regardless of what other threads do
* misaligned atomics should be disallowed by the language - they are UB in C++, for example
* WC memory can be supported by the language, with a requirement that the programmer insert an explicit serialisation operation

**Comment by RalfJung** [comment]
Replying to some [comments](https://github.com/rust-lang/unsafe-code-guidelines/issues/607#issuecomment-4152300922) by @comex that would be off-topic in #607:

>> mixed-size accesses are UB, so the untrusted code can still cause UB in the Rust code.
>
> But note that this UB could theoretically go away in the future. 

Yes, but it needs *way* more than just Intel clarifying. It needs completely re-doing the C++ memory model in a way that works with mixed-size accesses. Given the pace at which that model has been evolving, I would be extremely surprised if anything like that happened in the next 5 years, and would expect it to make more on the order of 10 years.

> Alternatively, we could continue to assume mixed-size atomics might tear, but make tearing well-defined, a concept which is desired anyway for atomic per-byte memcpy (as discussed in https://github.com/rust-lang/rfcs/pull/3301).

Note that atomic bytewise memcpy as specified in C++ says (in my understanding) that it is UB for such a memcpy to race with a non-memcpy access, and it is also UB for it to race with another memcpy that overlaps unless both access the exact same memory range. IOW, it makes no progress on the mixed-size access side.

**Comment by comex** [comment]
> Given the pace at which that model has been evolving, I would be extremely surprised if anything like that happened in the next 5 years, and would expect it to make more on the order of 10 years.

Agreed!  But I think those future prospects are still relevant to the question of which model (volatile or atomics) should be used when sharing memory with untrusted parties, which is why I brought this up in that thread.

> Note that atomic bytewise memcpy as specified in C++ says (in my understanding) that it is UB for such a memcpy to race with a non-memcpy access, and it is also UB for it to race with another memcpy that overlaps unless both access the exact same memory range. IOW, it makes no progress on the mixed-size access side.

Correct.  I was only talking about what is desired in connection with adding that feature to Rust, not what is allowed by the version that got into the C++ concurrency TS.

That said, skimming over https://github.com/rust-lang/rfcs/pull/3301 again this morning, I see that the desire to allow mixed-size accesses mostly came up in connection with interacting with untrusted code.  When I checked it it last night, I thought the desire was broader.  Personally I think it's just a natural fit.  If atomic per-byte memcpy were allowed to race with non-per-byte atomics, it would make it easier to design safe APIs around atomic per-byte memcpy, and would better fit programmers' mental models.

**Comment by chorman0773** [comment]
FYI, as part of redoing lccc's core infrastructure, I am going to be rewriting its memory model specification (in hopefully something more formal than just prose now). I'm going to attempt to cover mixed-sized (as well as nontemporal) accesses in the process, so it's possible we may have *a* model. Whether or not rust will want to adopt that model will, of course, be a separete question (as will wether or not llvm will actually adopt/obey it).

**Comment by vrtgs** [comment]
I agree with the comments here so far. Interacting with untrusted code that may concurrently read from or write to shared memory across multiple threads is a real challenge without mixed-size atomics, which is a large part of why they seem so valuable here. For example, in an ARM CPU emulator I have been working on, this has been a persistent pain point. One approach is to represent memory as a large slab of AtomicU8s and implement wider loads by reading each byte with Relaxed ordering and then reconstructing the value, with stores handled similarly. But that costs both performance and atomicity. In practice, this leaves me writing platform-specific inline assembly intrinsics so I can get mixed atomics for relaxed loads at each relevant bit width, which has been so far just, the basic CPU mov/(ld/st), which is exactly the sort of complexity I would love to avoid.

**Comment by RalfJung** [comment]
> One approach is to represent memory as a large slab of AtomicU8s and implement wider loads by reading each byte with Relaxed ordering 

This doesn't avoid the problem -- it's still UB to do such a 1-byte atomic load concurrently with an overlapping many-byte atomic store.

**Comment by vrtgs** [comment]
> > One approach is to represent memory as a large slab of AtomicU8s and implement wider loads by reading each byte with Relaxed ordering 
> 
> This doesn't avoid the problem -- it's still UB to do such a 1-byte atomic load concurrently with an overlapping many-byte atomic store.

I mean, you are the expert here so there is a high chance I am wrong, but the way I rationalized this is basically as follows: the assembly code just does `n` atomic loads and `n` atomic stores, they just happen to not get interrupted byt anything all the time since the assembly block is invisible to rust, as long as there is some series of transformations that can be made to the abstract virtual machine that can take the previous state of the machine before executing the assembly block, then applied and that produces a valid VM state afterwards its fine



And as for the assembly part, well ARM Single-copy atomicity, so if we take it as an example, its legal in the arch to fake the `n` byte stores as a series of stores that just happen to not get interrupted

**Comment by RalfJung** [comment]
I can't tell if that works, there are too many things I don't know about ARM concurrency.

**Comment by chorman0773** [comment]
This only works if you don't need a guaranteed total order of stores between those bytes (only independant total orders in each byte)

**Comment by vrtgs** [comment]
> This only works if you don't need a guaranteed total order of stores between those bytes (only independant total orders in each byte)

yea I don't, ARM has a pretty loose memory model, so for syncronizing it would use ldrex and strex or a memory barrier, and that WOULD give order

**Comment by robofinch** [comment]
Besides concerns about mixed-size atomic accesses (in the CM) that are also modeled in the AM as being mixed size, is anyone aware of hardware-level problems with angelically modeling (CM-level) mixed-size atomic accesses as (AM-level) relaxed *non*-mixed-size atomic accesses?

For atomic accesses on memory shared with untrusted parties, we can consider the untrusted party to either cooperate and avoid mixed-size accesses entirely, or be uncooperative in which case we don’t care whatsoever about which values are read from the shared memory or what happens-before relationships are established.

“*Surely* any problems (at the CM level) with mixed-size atomic accesses merely cause tearing and/or fewer established happens-before relationships on *any reasonable architecture*, so *surely* they should be fine (at the CM level) if you don’t care whatsoever about which values you read or what orderings are established,” is what I want to believe, but I don’t know enough about assembly-level concerns to justify those “surely”s or “any reasonable architecture”.

**Comment by RalfJung** [comment]
It seems you are cross-posting from https://github.com/rust-lang/unsafe-code-guidelines/issues/607. Please don't, let's keep the issues on-topic. This issue is strictly about AM-level mixed-sized accesses.
